File: stacks/largest_rectangle_in_histogram/largest_rectangle_in_histogram_test.go
package largestrectangleinhistogram

import "testing"

// TestLargestRectangleArea tests the largestRectangleArea function.
func TestLargestRectangleArea(t *testing.T) {
	// Define the test cases
	tests := []struct {
		name     string
		heights  []int
		expected int
	}{
		{
			name:     "Example 1 from problem description",
			heights:  []int{7, 1, 7, 2, 2, 4},
			expected: 8,
		},
		{
			name:     "Example 2 from problem description",
			heights:  []int{1, 3, 7},
			expected: 7,
		},
		{
			name:     "Single element",
			heights:  []int{5},
			expected: 5,
		},
		{
			name:     "All same heights",
			heights:  []int{2, 2, 2, 2},
			expected: 8,
		},
		{
			name:     "Ascending heights",
			heights:  []int{1, 2, 3, 4},
			expected: 6, // 3 * 2 = 6
		},
		{
			name:     "Descending heights",
			heights:  []int{4, 3, 2, 1},
			expected: 6, // 3 * 2 = 6
		},
	}

	// Iterate over all test cases
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := largestRectangleArea(tt.heights)

			if result != tt.expected {
				t.Errorf("got %d, want %d for heights %v", result, tt.expected, tt.heights)
			}
		})
	}
}

File: stacks/largest_rectangle_in_histogram/largest_rectangle_in_histogram_test.py
import pytest

# Assuming your function is named largestRectangleArea in the .py file
from largest_rectangle_in_histogram import largestRectangleArea


# Define all test cases in a clean, parameterized format
@pytest.mark.parametrize(
    "name, heights, expected",
    [
        ("example_1", [7, 1, 7, 2, 2, 4], 8),
        ("example_2", [1, 3, 7], 7),
        ("single_element", [5], 5),
        ("all_same_heights", [2, 2, 2, 2], 8),
        ("ascending_heights", [1, 2, 3, 4], 6),
        ("descending_heights", [4, 3, 2, 1], 6),
        ("empty_array", [], 0),
    ],
)
def test_largest_rectangle_area(name, heights, expected):
    """
    Tests the largestRectangleArea function with various edge cases.
    The 'name' parameter helps identify which specific test failed in the output.
    """
    assert largestRectangleArea(heights) == expected
